torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp

```diff
@@ -745,6 +873,49 @@ void propagateQuantizationOps(Block* block) {
   }
 }
 
+void InsertQuantDeQuantHelper::runWeightObserver(
+    Module& module,
+    const std::string& method_name) {
+  for (auto& invoked_methods : getInvokedMethods(module, method_name)) {
+    auto& invoked_module = std::get<0>(invoked_methods);
+    const auto& invoked_method_name = std::get<1>(invoked_methods);
+    runWeightObserver(invoked_module, invoked_method_name);
+  }
+  Method method = module.get_method(method_name);
+  auto graph = method.graph();
+  Value* self = graph->inputs()[0];
+
+  std::vector<Value*> weight_values;
+  // Visit all blocks in the current graph to find weight values.
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (auto n : b->nodes()) {
+      for (Value* v : n->outputs()) {
+        if (!v->type()->isSubtypeOf(TensorType::get())) {
+          continue;
+        }
+        auto observer_name = findObserverName(v);
+        if (observer_name && isWeight(module, v)) {
+          weight_values.push_back(v);
+        }
+      }
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+    }
+  }
+  // For all the observed weight values, find the corresponding subgraph that
+  // contributes to the weight tensor, and run that subgraph to observe the
+  // weight.
+  for (const auto& v : weight_values) {
+    ProcessWeightObserverSubgraph o;
+    o.extractAndRunWeightObserver(module, self, v->node());
```

Comment:
I think it's better to cache the extracted graph so that we don't need to extract the same graph multiple times

Reply:
I feel it's better here to just use `v` as input, since we'll extract subgraph for this weight value.